addons/controllers/cpi/cpiconfig_utils.go

```diff
@@ -0,0 +1,307 @@
+		}
+		if c.NSXT.ClientCertData != "" {
+			d.VSphereCPI.Nsxt.ClientCertData = c.NSXT.ClientCertData
+		}
+		if c.NSXT.RootCAData != "" {
+			d.VSphereCPI.Nsxt.RootCAData = c.NSXT.RootCAData
+		}
+		if c.NSXT.SecretName != "" {
+			d.VSphereCPI.Nsxt.SecretName = c.NSXT.SecretName
+			d.VSphereCPI.Nsxt.SecretNamespace = c.NSXT.SecretNamespace
+		}
+	}
+	if c.IPFamily != "" {
+		d.VSphereCPI.IPFamily = c.IPFamily
+	}
+	if c.HTTPProxy != "" {
+		d.VSphereCPI.HTTPProxy = c.HTTPProxy
+	}
+	if c.HTTPSProxy != "" {
+		d.VSphereCPI.HTTPSProxy = c.HTTPSProxy
+	}
+	if c.NoProxy != "" {
+		d.VSphereCPI.NoProxy = c.NoProxy
+	}
+	return d, nil
+}
+
+// mapCPIConfigToDataValuesParavirtual generates CPI data values for paravirtual modes
+func (r *CPIConfigReconciler) mapCPIConfigToDataValuesParavirtual(_ context.Context, cpiConfig *cpiv1alpha1.CPIConfig, _ *clusterapiv1beta1.Cluster) (*CPIDataValues, error) {
+	d := &CPIDataValues{}
+	c := cpiConfig.Spec.VSphereCPI
+
+	d.VSphereCPI.Mode = CPIParavirtualMode
+	d.VSphereCPI.ClusterAPIVersion = c.ClusterAPIVersion
+	d.VSphereCPI.ClusterKind = c.ClusterKind
+	d.VSphereCPI.ClusterName = c.ClusterName
+	d.VSphereCPI.ClusterUID = c.ClusterUID
+	d.VSphereCPI.SupervisorMasterEndpointIP = c.SupervisorMasterEndpointIP
+	d.VSphereCPI.SupervisorMasterPort = c.SupervisorMasterPort
+
+	return d, nil
+}
+
+// mapCPIConfigToDataValues maps CPIConfig CR to data values
+func (r *CPIConfigReconciler) mapCPIConfigToDataValues(ctx context.Context, cpiConfig *cpiv1alpha1.CPIConfig, cluster *clusterapiv1beta1.Cluster) (*CPIDataValues, error) {
+	switch cpiConfig.Spec.VSphereCPI.Mode {
+	case CPINonParavirtualMode:
+		return r.mapCPIConfigToDataValuesNonParavirtual(ctx, cpiConfig, cluster)
+	case CPIParavirtualMode:
+		return r.mapCPIConfigToDataValuesParavirtual(ctx, cpiConfig, cluster)
+	default:
+		break
+	}
+	return nil, errors.Errorf("Invalid CPI mode %s, must either be %s or %s", cpiConfig.Spec.VSphereCPI.Mode, CPIParavirtualMode, CPINonParavirtualMode)
+}
+
+// getOwnerCluster verifies that the CPIConfig has a cluster as its owner reference,
+// and returns the cluster. It tries to read the cluster name from the CPIConfig's owner reference objects.
+// If not there, we assume the owner cluster and CPIConfig always has the same name.
+func (r *CPIConfigReconciler) getOwnerCluster(ctx context.Context, cpiConfig *cpiv1alpha1.CPIConfig) (*clusterapiv1beta1.Cluster, error) {
+	cluster := &clusterapiv1beta1.Cluster{}
+	clusterName := cpiConfig.Name
+
+	// retrieve the owner cluster for the CPIConfig object
+	for _, ownerRef := range cpiConfig.GetOwnerReferences() {
+		if strings.EqualFold(ownerRef.Kind, constants.ClusterKind) {
+			clusterName = ownerRef.Name
+			break
+		}
+	}
+	if err := r.Client.Get(ctx, types.NamespacedName{Namespace: cpiConfig.Namespace, Name: clusterName}, cluster); err != nil {
+		if apierrors.IsNotFound(err) {
+			r.Log.Info(fmt.Sprintf("Cluster resource '%s/%s' not found", cpiConfig.Namespace, clusterName))
```

Comment:
sure, print a log it instead if cluster is not found

Reply:
@HanFa I believe this is still not fixed. We should return ```nil``` instead of err when the cluster is ```NotFound```, cause we do not want to requeue again for reconciliation

Reply:
Thanks for having fixed